ddsketch/ddsketch.go

```diff
@@ -1,159 +1,108 @@
 // under the Apache License 2.0.
 // This product includes software developed at Datadog (https://www.datadoghq.com/).
-// Copyright 2018 Datadog, Inc.
+// Copyright 2020 Datadog, Inc.
 
 package ddsketch
 
 import (
-	"bytes"
-	"fmt"
-	"math"
-	"reflect"
+	"errors"
+
+	"github.com/DataDog/sketches-go/ddsketch/store"
 )
 
-// DDSketch is an implementation of DDSketch.
 type DDSketch struct {
-	config *Config
-	store  *Store
-	min    float64
-	max    float64
-	count  int64
-	sum    float64
+	IndexMapping
+	store     store.Store
+	zeroCount float64
 }
 
-// NewDDSketch allocates a new DDSketch summary with relative accuracy alpha.
-func NewDDSketch(c *Config) *DDSketch {
+func NewDDSketch(indexMapping IndexMapping, store store.Store) *DDSketch {
 	return &DDSketch{
-		config: c,
-		store:  NewStore(c.maxNumBins),
-		min:    math.Inf(1),
-		max:    math.Inf(-1),
+		IndexMapping: indexMapping,
+		store:        store,
 	}
 }
 
-// Add a new value to the summary.
-func (s *DDSketch) Add(v float64) {
-	key := s.config.Key(v)
-	s.store.Add(key)
-
-	// Keep track of summary stats
-	if v < s.min {
-		s.min = v
-	}
-	if s.max < v {
-		s.max = v
+func MemoryOptimalCollapsingLowestSketch(relativeAccuracy float64, maxNumBins int) (*DDSketch, error) {
+	indexMapping, err := NewLogarithmicMapping(relativeAccuracy)
+	if err != nil {
+		return nil, err
 	}
-	s.count++
-	s.sum += v
+	return NewDDSketch(indexMapping, store.NewCollapsingLowestDenseStore(maxNumBins)), nil
 }
 
-// Quantile returns the estimate of the element at q.
-func (s *DDSketch) Quantile(q float64) float64 {
-	if q < 0 || q > 1 || s.count == 0 {
-		return math.NaN()
-	}
-
-	if q == 0 {
-		return s.min
-	} else if q == 1 {
-		return s.max
+func (s *DDSketch) Accept(value float64) error {
+	if value < 0 || value > s.MaxIndexableValue() {
+		return errors.New("The input value is outside the range that is tracked by the sketch.")
 	}
 
-	rank := int(q*float64(s.count-1) + 1)
-	key := s.store.KeyAtRank(rank)
-	var quantile float64
-	if key < 0 {
-		key += s.config.offset
-		quantile = -2 * s.config.powGamma(-key) / (1 + s.config.gamma)
-	} else if key > 0 {
-		key -= s.config.offset
-		quantile = 2 * s.config.powGamma(key) / (1 + s.config.gamma)
+	if value < s.MinIndexableValue() {
+		s.zeroCount++
 	} else {
-		quantile = 0
+		s.store.Add(s.Index(value))
 	}
-	// Check that the returned value is larger than the minimum
-	// since for q close to 0 (key in the smallest bin) the midpoint
-	// of the bin boundaries could be smaller than the minimum
-	if quantile < s.min {
-		quantile = s.min
-	}
-	if quantile > s.max {
-		quantile = s.max
-	}
-	return quantile
+	return nil
 }
 
-// Merge another sketch (with the same maxNumBins and gamma) in place.
-func (s *DDSketch) Merge(o *DDSketch) {
-	if o.count == 0 {
-		return
+func (s *DDSketch) getValueAtQuantile(quantile float64) (float64, error) {
+	if quantile < 0 || quantile > 1 {
+		return 0, errors.New("The quantile must be between 0 and 1.")
```

Comment:
Basic Go question: from the user's point of view, when this happens does it both return 0 and throw an error? I'm wondering about this for JS implementation too (whether to fail loudly or try to give some answer), and it's probably worth trying to be consistent for all of our languages?

Reply:
It will return both 0 and an error - the user needs to check whether the error is nil before using the value. `math.NaN` is probably a better value to return in this case rather than 0.